independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Methods.java

```diff
@@ -216,16 +217,27 @@ static Set<MethodInfo> addInterceptedMethodCandidates(BeanDeployment beanDeploym
                 //interfaces can't have final methods
                 addInterceptedMethodCandidates(beanDeployment, interfaceInfo, originalClassInfo, candidates,
                         classLevelBindings, bytecodeTransformerConsumer, transformUnproxyableClasses,
-                        skipPredicate, true);
+                        skipPredicate, true, noClassInterceptorsMethods);
             }
         }
         return finalMethodsFoundAndNotChanged;
     }
 
     private static Set<AnnotationInstance> mergeBindings(BeanDeployment beanDeployment, ClassInfo classInfo,
-            Set<AnnotationInstance> classLevelBindings, boolean ignoreMethodLevelBindings, MethodInfo method) {
-
-        if (beanDeployment.getAnnotation(method, DotNames.NO_CLASS_INTERCEPTORS) != null) {
+            Set<AnnotationInstance> classLevelBindings, boolean ignoreMethodLevelBindings, MethodInfo method,
+            Set<MethodKey> noClassInterceptorsMethods) {
+
+        MethodKey key = new MethodKey(method);
+        if (beanDeployment.getAnnotation(method, DotNames.NO_CLASS_INTERCEPTORS) != null
+                || noClassInterceptorsMethods.contains(key)) {
+            // The set of methods with `@NoClassInterceptors` is shared in the traversal of class hierarchy,
+            // so once a method with the annotation is found, all subsequent occurences of the "same" method
+            // are treated as if they also had it. Given that we traverse classes bottom-up, this works as
+            // expected (presence of `@NoClassInterceptors` on subclass overrides absence on superclass).
+            // We process interfaces after all superclasses, but we inspect interfaces implemented by classes
+            // in a top-down manner, so `@NoClassInterceptors` on interfaces may behave weirdly. It shouldn't
```

Comment:
Would it make sense to detect `@NoClassInterceptors` on interfaces and throw an exception?

Reply:
Thinking about it again, I can see a use case for `default` method on interfaces (and it would work correctly if such method wasn't overridden), so I'd be a little cautious.

We could detect it on abstract methods on interfaces, probably.

Reply:
Depends how exactly we need this to work.
In a true CDI manner, you should probably ignore any annotation on interfaces and avoid carrying in over to bean definition (an interceptor binding wouldn't carry over from interface as well). However, that might block too many use cases for this annotation? TBF I don't know what exactly are we using this for so maybe I am just creating noise here :-)

Reply:
That is a good point, but we already do apply interceptors to default methods on interfaces. I'll think a bit more about this.

Reply:
> but we already do apply interceptors to default methods on interfaces

Hm, that's true and we only apply class-level ones, right? I mean, class-level annotation from a bean applies to inherited default interface method. So technically, there is no other way to prevent that other than this annotation.
With that in mind, you're right that we could only ignore abstract interface methods.
I'll also give it some more thought

Reply:
I took a look at this and added a test and adjusted the comment. Basically, this works fine as is even for `default` methods, when they are inherited only once. The current code only behaves weirdly when "the same" `default` method is inherited multiple times. IMHO, we can just leave that be and get back to it when it becomes problematic.